src/passes/J2CLItableMerging.cpp

```diff
@@ -0,0 +1,347 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This is a J2CL specific pass that merges itables into vtables. It is meant
+// to be run at the beginning before structs corresponding to Java classes are
+// optimized.
+//
+// The motivation for embedding itables into vtables is to reduce memory usage.
```

Comment:
The effects are as follows: 

  - on the heap, each "Java"  object has on field less, so this affects every instance of Java objects. 
  - on instance creation there is one less `global.get`.
  - there are fewer globals since the 1 per class itable globals is now removed. This also has the impact of having smaller global indices.
  - in some cases fewer `struct.get`s since if there are vtable and itable dispatches on the same instance there is only one needed instead of two.

Reply:
Ah, thanks, especially removing a field on every single Java object sounds very useful!